docs: Document PostgreSQL database support for Mautic 7.x (#310)
mautic/developer-documentation-new: docs/plugins/database.rst

        :param array $columns: Array of columns to included in the index.

        :return: ``INDEX {tableName} ($columns...)`` statement
        :returntype: string

> Database compatibility

Author: Database compatibility section based on [PR #15926](https://github.com/mautic/mautic/pull/15926). Key implementation files: `CommonRepository.php`, `GeneratedColumnProvider.php`.

Source: https://github.com/mautic/mautic/pull/15926

Supported databases
===================

.. list-table::
    :widths: 30 30 40
    :header-rows: 1

    * - Database
      - Minimum version
      - PHP extension
    * - MySQL
      - 8.0
      - ``pdo_mysql``
    * - MariaDB
      - 10.6
      - ``pdo_mysql``
    * - PostgreSQL
      - 16
      - ``pdo_pgsql``

.. vale off

Case-insensitive string matching
================================

.. vale on

MySQL and MariaDB use case-insensitive ``LIKE`` comparisons by default. PostgreSQL's ``LIKE`` is case-sensitive. Mautic provides helper methods in ``CommonRepository`` to handle this.

**Using the helper methods:**

.. code-block:: php

    // In a repository extending CommonRepository
    $qb = $this->getEntityManager()->getConnection()->createQueryBuilder();

    // Case-insensitive LIKE - uses ILIKE on PostgreSQL, LIKE on MySQL/MariaDB
    $qb->andWhere(
        $this->getILikeExpression($qb, 'l.email', ':search')
    );

    // Or wrap the column with LOWER() for case-insensitive comparison
    $qb->andWhere(
        $this->getLowerLikeExpression($qb, 'l.firstname', ':search')
    );

**Available methods in CommonRepository:**

> - ``getILikeExpression(QueryBuilder $qb, string $column, string $parameter)``: Returns platform-appropriate case-insensitive ``LIKE`` expression

Reviewer: ```suggestion
* ``getILikeExpression(QueryBuilder $qb, string $column, string $parameter)``: returns platform-appropriate case-insensitive ``LIKE`` expression
```

> - ``getLowerLikeExpression(QueryBuilder $qb, string $column, string $parameter)``: Wraps the column with ``LOWER()`` for case-insensitive comparison

Reviewer: ```suggestion
* ``getLowerLikeExpression(QueryBuilder $qb, string $column, string $parameter)``: wraps the column with ``LOWER()`` for case-insensitive comparison
```

- ``isPostgreSql()``: Returns ``TRUE`` if connected to a PostgreSQL database

.. vale off

Column and alias quoting
========================

.. vale on

PostgreSQL lowercases unquoted identifiers automatically. This causes issues with Mautic's ``camelCase`` column aliases. Always quote identifiers in raw SQL queries that use mixed-case names.

> **Use quoted identifiers for camelCase aliases:**

Reviewer: ```suggestion
.. vale off

**Use quoted identifiers for camelCase aliases:**

.. vale on
```

.. code-block:: php

    // Correct - aliases are quoted
    $qb->select('l.id, l.first_name AS "firstName", l.date_added AS "dateAdded"');

    // Incorrect - PostgreSQL converts to lowercase
    $qb->select('l.id, l.first_name AS firstName');

Doctrine's QueryBuilder handles quoting automatically when you use proper field mappings. You only need manual quoting for raw SQL or custom column aliases.

.. vale off

GROUP BY requirements
=====================

.. vale on

> PostgreSQL enforces strict ``GROUP BY`` rules, and MySQL 8+ does the same when ``ONLY_FULL_GROUP_BY`` SQL mode is enabled - the default in strict mode. Every column in the ``SELECT`` clause must appear in the ``GROUP BY`` clause or use an aggregate function.

Reviewer: Use active instead of passive voice. See "is enabled".

    $qb->select('l.id, l.email, COUNT(e.id) as emailCount')
       ->from('leads', 'l')
       ->leftJoin('l', 'emails', 'e', 'e.lead_id = l.id')
       ->groupBy('l.id');

Detecting the database platform
===============================

When you need platform-specific query logic, detect the database type using the connection's platform:

.. code-block:: php

    use Doctrine\DBAL\Platforms\PostgreSQLPlatform;

    $connection = $this->getEntityManager()->getConnection();
    $platform = $connection->getDatabasePlatform();

    if ($platform instanceof PostgreSQLPlatform) {
        // PostgreSQL-specific logic
    } else {
        // MySQL/MariaDB logic
    }

Repositories extending ``CommonRepository`` can use the ``isPostgreSql()`` helper method.

Best practices
==============

1. **Use Doctrine's ORM and QueryBuilder** - Doctrine abstracts most database differences. Avoid raw SQL when possible.

2. **Test on multiple databases** - Mautic's CI tests against MySQL, MariaDB, and PostgreSQL. Run your Plugin tests against all platforms before release.

3. **Quote mixed-case aliases** - When using custom column aliases with ``camelCase`` names in raw SQL, always quote them.

> 4. **Use repository helper methods** - ``CommonRepository`` provides cross-platform helpers for common operations like case-insensitive searches.

Reviewer: Follow our style guide for ordered list.

```suggestion
#. **Use Doctrine's ORM and QueryBuilder** - Doctrine abstracts most database differences. Avoid raw SQL when possible.
#. **Test on multiple databases** - Mautic's CI tests against MySQL, MariaDB, and PostgreSQL. Run your Plugin tests against all platforms before release.
#. **Quote mixed-case aliases** - When using custom column aliases with ``camelCase`` names in raw SQL, always quote them.
#. **Use repository helper methods** - ``CommonRepository`` provides cross-platform helpers for common operations like case-insensitive searches.
```

    * - PostgreSQL
      - 16
      - ``pdo_pgsql``

.. vale off

Case-insensitive string matching
================================

.. vale on

MySQL and MariaDB use case-insensitive ``LIKE`` comparisons by default. PostgreSQL's ``LIKE`` is case-sensitive. Mautic provides helper methods in ``CommonRepository`` to handle this.

**Using the helper methods:**

.. code-block:: php

    // In a repository extending CommonRepository
    $qb = $this->getEntityManager()->getConnection()->createQueryBuilder();

    // Case-insensitive LIKE - uses ILIKE on PostgreSQL, LIKE on MySQL/MariaDB
    $qb->andWhere(
        $this->getILikeExpression($qb, 'l.email', ':search')
    );

    // Or wrap the column with LOWER() for case-insensitive comparison
    $qb->andWhere(
        $this->getLowerLikeExpression($qb, 'l.firstname', ':search')
    );

**Available methods in CommonRepository:**

* ``getILikeExpression(QueryBuilder $qb, string $column, string $parameter)``: returns platform-appropriate case-insensitive ``LIKE`` expression
* ``getLowerLikeExpression(QueryBuilder $qb, string $column, string $parameter)``: wraps the column with ``LOWER()`` for case-insensitive comparison

> - ``isPostgreSql()``: Returns ``TRUE`` if connected to a PostgreSQL database

Reviewer: ```suggestion
* ``isPostgreSql()``: returns ``TRUE`` if connected to a PostgreSQL database
```

**Correct pattern:**

.. code-block:: php

    $qb->select('l.id, l.email, COUNT(e.id) as emailCount')
       ->from('leads', 'l')
       ->leftJoin('l', 'emails', 'e', 'e.lead_id = l.id')
       ->groupBy('l.id, l.email');  // All non-aggregate columns included

**Incorrect pattern:**

.. code-block:: php

    // This fails on PostgreSQL - l.email not in GROUP BY
    $qb->select('l.id, l.email, COUNT(e.id) as emailCount')
       ->from('leads', 'l')
       ->leftJoin('l', 'emails', 'e', 'e.lead_id = l.id')
       ->groupBy('l.id');

Detecting the database platform
===============================

When you need platform-specific query logic, detect the database type using the connection's platform:

.. code-block:: php

    use Doctrine\DBAL\Platforms\PostgreSQLPlatform;

    $connection = $this->getEntityManager()->getConnection();
    $platform = $connection->getDatabasePlatform();

    if ($platform instanceof PostgreSQLPlatform) {
        // PostgreSQL-specific logic
    } else {
        // MySQL/MariaDB logic
    }

Repositories extending ``CommonRepository`` can use the ``isPostgreSql()`` helper method.

Best practices
==============

> #. **Use Doctrine's ORM and QueryBuilder** - Doctrine abstracts most database differences. Avoid raw SQL when possible.

Reviewer: There are a couple of things:

- Expand ORM
- Add ORM to `accept.txt` file

```suggestion
#. **Use Doctrine's Object Relational Mapper - ORM - and QueryBuilder** - Doctrine abstracts most database differences. Avoid raw SQL when possible.
```
